### PR TITLE
Change of range in product - fix 2037066

### PR DESCRIPTION
**Pull request by mtygesen**
Fixes:
https://bugs.launchpad.net/tapaal/+bug/2037066

+ Fixed bug where the arc was not updated when changing integer range
+ Propogated integer range to transition guard
+ Added error message when trying to add a new color type with a similar integer range as an existing type

### CODE REVIEW

File: src/main/java/net/tapaal/gui/petrinet/undo/Colored/UpdateColorTypeCommand.java [hunk 1]
```diff
@@ -75,19 +82,27 @@ public void redo() {
...
+                    transition.getGuard().setColorType(newColorType);
+                }
             }
         }
 
         eval(newColorType);
 
         for (Variable var : network.variables()) {
             if (var.getColorType().equals(oldColorType)) {
+                System.out.println();
```
Comment (srba) [inline src/main/java/net/tapaal/gui/petrinet/undo/Colored/UpdateColorTypeCommand.java:105]: This should be removed from the PR

File: src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java [hunk 2]
```diff
@@ -684,6 +684,16 @@ public Integer getVariableIndex(String name) {
         return null;
     }
 
+	public boolean doesSimilarRangeExist(ColorType newColorType) {
```
Comment (srba) [inline src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java:687]: Please, rename from ...SimilarRange... to ...Identical range...

File: src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java [hunk 3]
```diff
@@ -684,6 +684,16 @@ public Integer getVariableIndex(String name) {
...
+		for (ColorType ct : colorTypes) {
+			if (ct.getColors().toString().equals(newColorType.getColors().toString()) && !ct.equals(newColorType)) {
```
Comment (srba) [inline src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java:689]: Is not there a better way to check that the integer ranges are the same, without converting to string comparison?

File: src/main/java/net/tapaal/gui/petrinet/editor/ColorTypeDialogPanel.java [hunk 4]
```diff
@@ -1078,6 +1079,14 @@ private void onOK() {
                     newColorType.addColor(String.valueOf(i));
                 }
 
+                if (network.isIndeticalToExisting(newColorType)) {
+                    JOptionPane.showMessageDialog(
+                        TAPAALGUI.getApp(),
+                        "Color type with similar range already exists!",
```
Comment (yrke) [inline src/main/java/net/tapaal/gui/petrinet/editor/ColorTypeDialogPanel.java:1085]: Same comment Similar to Identical

File: src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java (hunk 3 above)
Comment (yrke) [reply src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java:689]: If there is not a way to get the semantic value we should add it.

File: src/main/java/dk/aau/cs/model/CPN/ColorType.java [hunk 5]
```diff
@@ -40,10 +40,14 @@ public Integer size() {
         return colors.iterator();
     }
 
-    public Vector<Color> getColors(){
+    public Vector<Color> getColors() {
         return colors;
     }
 
+    public String getSemanticValue() {
```
Comment (srba) [inline src/main/java/dk/aau/cs/model/CPN/ColorType.java:47]: Well, this is actually the same as before (still comparing strings) and in fact I would prefer not to have this conversion to string as a new method doing just .toString(); anyway.  What I had in mind was whether one could implement a method returning the range (as a pair of integers) and compare these numbers to check if an identical integer range already exists. But I am not sure if this is easy to implement, if not than the (original) string comparison method could be also acceptable.

File: src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java [hunk 6]
```diff
@@ -686,7 +686,7 @@ public Integer getVariableIndex(String name) {
 	
 	public boolean isIndeticalToExisting(ColorType newColorType) {
 		for (ColorType ct : colorTypes) {
-			if (ct.getSemanticValue().equals(newColorType.getSemanticValue()) && !ct.equals(newColorType)) {
+			if (ct.isIdentical(newColorType) && !ct.equals(newColorType)) {
```
Comment (yrke) [inline src/main/java/dk/aau/cs/model/tapn/TimedArcPetriNetNetwork.java:689]: nice, I think you could also add the equals check to the isIdentical, so comparing with once self is also handled there :)

**Review by yrke — APPROVED** [review]

File: src/main/java/dk/aau/cs/model/CPN/ColorType.java [hunk 7]
```diff
@@ -40,10 +40,18 @@ public Integer size() {
...
```
Comment (srba) [inline src/main/java/dk/aau/cs/model/CPN/ColorType.java:47]: Please, remove this method, it is never used as far as I can see.

File: src/main/java/dk/aau/cs/model/CPN/ColorType.java [hunk 8]
```diff
@@ -89,6 +97,18 @@ public Color getFirstColor() {
...
+    }
+
+    private String getFirstColorName() {
+        return getFirstColor().getColorName();
+    }
+
+    private String getLastColorName() {
+        return getLastColor().getColorName();
+    }
+
```
Comment (srba) [inline src/main/java/dk/aau/cs/model/CPN/ColorType.java:111]: I would suggest not to add these methods and move this directly to the if-condition where it is used.

**Review by srba — APPROVED** [review]
Fixes all the problems.

**Pull request merged**